pulp_2to3_migration/app/pre_migration.py

```diff
@@ -621,14 +635,23 @@ def delete_old_resources(plan):
     old_dist_query &= Q(pulp2_repository__in=repos_with_old_distributions_qs) | Q(not_in_plan=True)
 
     with transaction.atomic():
-        pulp2distributors_with_old_distributions_qs = Pulp2Distributor.objects.filter(
-            old_dist_query).only('pulp3_distribution')
-
+        pulp2distributors_with_old_distributions_qs = Pulp2Distributor.objects.filter(old_dist_query)
+        pb_to_delete = set()
+        dist_to_delete = []
         for pulp2distributor in pulp2distributors_with_old_distributions_qs:
             if pulp2distributor.is_migrated:
                 pulp2distributor.is_migrated = False
                 pulp2distributor.save()
             if pulp2distributor.pulp3_publication:
-                pulp2distributor.pulp3_publication.delete()
+                # check if publication is shared by multiple distributions
+                pulp2dists = pulp2distributor.pulp3_publication.pulp2distributor_set.all()
+                publication = pulp2distributor.pulp3_publication
+                for dist in pulp2dists:
+                    if dist.is_migrated:
+                        dist.is_migrated = False
+                        dist.save()
```

Comment:
Why is it enough to mark distributors as `is_migrated=False` and not save their pk for removal?

Reply:
distributions that have been affected indirectly ( aka they lost their publication only) it is enough just to flip the is_migrated flag to False so then the distribution will be updated with the new publication.
distributions that have been affected directly( aka repos were recreated in pulp2), will be removed , because the corresponding pulp2distributor will be present in this `pulp2distributors_with_old_distributions_q`

Reply:
interesting that the affected ones are updated later

Reply:
this line is hit where we take into consideration distributors that are in plan and is_mifrated=False https://github.com/pulp/pulp-2to3-migration/blob/master/pulp_2to3_migration/app/migration.py#L224
then we are trying to migrate that distributor https://github.com/pulp/pulp-2to3-migration/blob/master/pulp_2to3_migration/app/migration.py#L331
and here we are 1) taking already created publication from the repo and update the existing distribution https://github.com/pulp/pulp-2to3-migration/blob/master/pulp_2to3_migration/app/plugin/iso/repository.py#L76 that why it is unnecessary to remove it in the remove_old_resources

Reply:
Thanks!  The missing part for me was that we take care of publication update in every plugin which has publications. And also publication ends up as a part of `base_config`, it was unexpected I missed that.

So I'm torn now. Let me know what you think.
My concern is that plugin writer needs to know that they are responsible for the update case (when publication changed) and not only for a create one. So we can either add some docs to the plugin writer's guide about that. Or we can remove distributors here and they will only worry about creation.
I'm fine with both, not so many plgin writers will need to extend the migration plugin. Just the problem is painful if the implementation of update case is missed.

Reply:
if we keep publication or repo_version( depending on the plugin) among base_config data then it will be handled in obth cases, whether it is a create or update or the object.
If we remove the distribution but this line is missing https://github.com/pulp/pulp-2to3-migration/blob/master/pulp_2to3_migration/app/plugin/iso/repository.py#L75 it does not really matter whether we remove the distribution or just flip the flag in pulp2dist to later update the distribution, in both cases distribution object will miss the publication

Reply:
tldr the problem is not in the update or creation of the object but to make sure we include the relation between the publication and distribution

Reply:
+1 to document that it is important to use update_or_create in distributions/remotes